### ISSUE TITLE
TC Trap phasing is (probably?) incorrect

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#629]
I discovered something interesting while writing tests for Borealis that I'm tentatively (and hesitantly) calling an actual bug in the AGC hardware.

During my test of interrupt priorities and EDRUPT's ability to trigger interrupt routines, I need to wait for TIME3 to increment so the phasing of the timers is correct for the remainder of the test. Ed Smally had to do something very similar for his RUPTCHK:

```
                CA              S+ZERO                          
                TS              ZRUPT                           
                RELINT                                          
                AD              TIME4                           
                TS              SKEEP1                          
TENMS           CS              SKEEP1                          
                AD              TIME4                           # WAIT FOR NEXT TIME4 INCREMENT
                EXTEND                                          
                BZF             TENMS
```

Basically, he captures the value of TIME4 and then waits for it to change, using a subtraction and a BZF. This works absolutely fine on both VirtualAGC and the hardware sim. I found, though, that when I used _the exact same structure_ to wait for TIME3, I was occasionally getting TC Traps on VirtualAGC for not using any TCs or TCFs for too long. Without thinking too much about it, I inserted a NOOP (which assembles to TCF +1) and moved on.

Thinking about it later, though, I've become progressively more confused about why Ed didn't need any sort of TC/TCF in his 10ms loop. The documentation-stated behavior of the TC Trap alarm (which is backed up by the schematics) is that the time it takes to trigger a TC Trap varies from 5ms to 15ms -- so it certainly seems like his accurately-labeled TENMS loop should run into trouble here... and yet, it doesn't.

It's worth noting that EXTEND, despite really being a "TC 6", does NOT count for the TC Trap, as a TC never gets executed on the real hardware; I can back these claims up with the schematics, but basically what happens during an extend is as follows. The instruction code for the next instruction is completed being read in from memory during timepulse 7 of the last MCT of its preceding instruction. The parity pyramid, as a side effect, notices if the word loaded is exactly 00006, and if so (and it's phase 4 of timepulse 7), the extend flip-flop is immediately set. The next instruction sequence executed, then, is not TC, but STD2 (the standard stage 2 sequence, which simply takes care of loading the next instruction).

So what's going on? Why does Ed's code work? For reference, here is the circuit for the TC Trap alarm:
![tctrap_circuit](https://cloud.githubusercontent.com/assets/94056/22053094/5fba6b34-dd03-11e6-9f66-aea27ac9417d.png)
It's divided into two halves -- the top half, which makes sure TC happens frequently enough, and the bottom half, which makes sure TC isn't the only instruction executed. The timing is based around pulses from the F10 scaler frequency, all of which have a period of 10ms -- and F10A and F10B are 5ms out of phase with each other. The basic idea is that at F10B, the flip-flop of each half is set. It is then up to the software to fulfill the requirements to _unset_ that flip-flop before F10A rolls around. If, when F10A happens 5ms later, either half still has its flip-flop set, then a TC Trap is declared and the computer is reset.

Right off the bat, there's two important things to notice about this circuit:
1. The requirements for the "instruction cannot be only TC/TCF" case are more precise than for the "no TC/TCF" case. The former specifically samples on timepulse 4, and by also checking INKL guarantees that any unprogrammed counter sequences don't interfere with the alarm (INKL is set during every counter sequence). The latter simply requires the TC0 or TCF0 sequences to be selected _at any point in time_. I'll come back to why this is important later.
2. The F10A _must_ be a schematic error. F10A is low the vast majority of the time, and if the polarity shown is correct, then a TC Trap would happen immediately on every F10B. There are no other derived frequencies of F10 which make sense here, so the best assumption to make is that either the schematic should read as F10A/, F10A's inverse, or a NOT gate was omitted from the schematic.

Because all of the timers are defined in terms of scaler frequencies, there's a well-established relationship between TIME4 increments and the F10A and F10B pulses. This relationship is extremely important for the issue at hand -- there are some possible relationships for which Ed's code is actually correct. For example, if TIME4 increments after F10B but before F10A, then his loop will _always_ exit before the TC Trap finishes looking. If TIME4 does _not_ increment between F10B and F10A, then the TC trap will almost certainly trigger.

Here is how the various timer pulses are constructed, with TIME4's increment logic circled:
![timer_logic](https://cloud.githubusercontent.com/assets/94056/22053632/ddf45b24-dd06-11e6-9ab9-070e659f3d93.png)

TIME4's increment pulse T4P is defined in terms of FS10 (the square-wave version of F10) and F09B. We can safely ignore the CNTRSB/ input; it is equivalent the inverse of SB2, one of the strobe pulses, and it happens so much more frequently than FS10 and F09B that it's irrelevant for this discussion. This circuit is more or less equivalent to "allow through every other F09B pulse as T4P" -- specifically the F09B pulse that occurs when FS10 is low. (F09B has a 5ms period, so allowing through every other gives the desired 10ms timer period). 

Let's step back and look at the relationship between all of these scaler frequencies. These are the circuits that define all of the F09 and F10 scaler frequency scalers:
![f09f10](https://cloud.githubusercontent.com/assets/94056/22054277/9ca7849a-dd09-11e6-9dc3-37200df80771.png)
(Note: I rearranged these and removed irrelevant circuits for clarity)

As can be seen, each stage of the scaler is a nifty little 6-gate circuit. For this application, it generates both polarities of a 50%-duty cycle square wave with half the frequency of the input clock, as well as two frequencies (FxxA and FxxB) with much lower duty cycle whose pulses are separated by 180 degrees. Each stage of the scaler is driven by the FxxA pulse line of the scaler stage preceding it. It's difficult to explain in words exactly how this works, so here's a gif of the two scaler stages in motion:
![scaler_freqs](https://cloud.githubusercontent.com/assets/94056/22054799/5a8f81a4-dd0c-11e6-819f-36095580e9d1.gif)

There's a few important relationships here:
1. F10A and F10B are always synchronous with F09A, and have the same duration as it. This holds true for every scaler stage: the FyyA and FyyB frequencies of any stage will always occur at the exact same time as the FxxA pulse of the previous stage, just alternating. Since F09A and F09B have a period of 5ms, they are separated by 2.5ms. And since F10A and F10B always align with F09A, there will always be an F09B pulse 2.5ms after each F10A and each F10B.
2. F10A always precedes FS10 going low. Likewise, F10B always precedes FS10 going high.

This gives us enough information to construct T4P and place it relative to the TC Trap sample window. Here are all of the relevant waveforms in the hardware sim, up to and including T4P:
![tctrap_waveforms](https://cloud.githubusercontent.com/assets/94056/22055142/77f19758-dd0e-11e6-91a8-07c179bae56a.jpg)
All relationships exhibited are as expected based on the above: F10A and F10B are synchronous with F09A. F10B immediately precedes FS10 going high. And TIME4 is equivalent to the F09B pulse that happens while FS10 is low.

Unfortunately, this places TIME4 incrementing squarely outside of the 5ms window during which TC Trap is active.

So what about VirtualAGC? Well, as it turns out, I got the phasing 180 degrees off when I implemented TC Traps. If I make it print out "S" when it starts sampling for TC Trap, "E" when it ends, and "4" when TIME4 is incremented, I get:
```
mstewart@archturus:~/virtualagc/yaAGC$ ./yaAGC --exec=../Aurora12/Aurora12.bin
Apollo Guidance Computer simulation, ver. 2016-11-08-working, built Jan 15 2017 14:44:16
Copyright (C) 2003-2009 Ronald S. Burkey, Onno Hommes.
yaAGC is free software, covered by the GNU General Public License, and you are
welcome to change it and/or distribute copies of it under certain conditions.
Refer to http://www.ibiblio.org/apollo/index.html for additional information.
(agc) run
[New Thread 11]
[Switching to Thread 11]
S4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4ES4E
```
I didn't pay much attention to the phasing of the alarms during implementation, under the apparently incorrect assumption that it didn't matter. And I've managed to implement it such that TIME4 increments in the middle of the TC Trap window. Here's what happens in Aurora self-tests if I correct the phasing:
![aurora_fail](https://cloud.githubusercontent.com/assets/94056/22055479/366bb35c-dd10-11e6-87da-cd39759fca65.gif)

Aurora's self-tests immediately trigger a TC Trap restart! (Channel 77 being 4 means the restart cause was a TC Trap).

So then, why does Aurora _pass_ on the simulated hardware, when it should clearly be provoking a TC Trap? Well, remember the other interesting thing about the "no TCs" half TC Trap circuit: it only requires that the TC0 or TCF0 instruction sequences be selected _at any time_, unlike the "only TCs" half which specifically examines timepulse 4. Here's the waveform view of Ed's exact loop executing in the hardware sim:
![transients](https://cloud.githubusercontent.com/assets/94056/22055888/2ab3a7f2-dd12-11e6-99de-b0507516ead1.jpg)

Due to propagation delays in the SQ register module, there's occasional transients on the TC0 and TCF0 lines during instruction changeover. They're short, and for the most part don't affect anything... but they _are_ long enough to reset the ungated TC Trap flip-flop. And so a TC Trap never happens.

So, that leaves us with an interesting predicament: the phasing of the TC Trap sample window is wrong in VirtualAGC, but the original test code doesn't pass if I fix it because it _shouldn't_ have worked in the first place.

**Comment by rburkey2005** [comment]
Assuming I read this right, I agree that a satisfactory resolution to this issue may not be possible.  How intent should we be on duplicating AGC bugs &mdash; if that's what that is &mdash; in Virtual AGC?  I find that I can argue it either way, equally convincingly, equally unsatisfyingly.

As far as yaAGC is concerned, perhaps there should be a command-line switch to determine whether yaAGC duplicates this bug or does not.  That would allow someone to choose between running a simulation "safely" vs. "authentically".  I guess you don't have that luxury with the hardware sim.

It does seem to me to be something to pursue with Hugh once your deliberations are complete, and possibly with Ramón Alonso if he's accessible, to see if they _agree_ that it's classifiable as an AGC bug.  If they do, then at very least it should be documented as part of the EDRUPT description on the website ... in a much more dumbed-down fashion, limited to just the circumstances in which EDRUPT might fail!  Obviously, the explanation can hyperlink to this issue for those desiring more info.

**Comment by thewonderidiot** [comment]
Yes, I'll definitely talk this over with Hugh when I get a chance -- and I'd love to talk to Ramón Alonso if he can be contacted.

My next course of action, I think, is going to be to manually trace through the instruction changeover logic on paper and see if I can derive a formula for when these transients occur... thus, at least, proving to myself that they're not artifacts of my simulation. If I can come up with expressions for the transients, I'd vote for correcting the phasing in VirtualAGC and "simulating" the transients by resetting the TC Trap when the expressions evaluate to true.

**Comment by thewonderidiot** [comment]
So I took a look at it, and it looks like these are the basic rules for the transients:
1. Slightly early clearing of extend for BZF leads to a TCF0 transient
2. TCF will cause a TC0 transient
3. INHINT or RELINT after single-cycle instruction lead to a TC0 transient
4. (The major one) If the next opcode to be executed is equal to TC or TCF if its most significant bit is zeroed, the respective line will get a transient
5. EXTEND being the next instruction cancels all of the above

I think there's other mechanisms, but these rules cover the vast majority of the cases I'm seeing. Still, not particularly straightforward.

**Comment by thewonderidiot** [comment]
As a triple-check that my understanding of the scaler and timer phasing was correct, I spent some time tonight breadboarding the F09 and F10 stages of the scaler, as well as the gates that construct T4P.
![t3t4](https://cloud.githubusercontent.com/assets/94056/22583907/5a43d386-e9a4-11e6-88b1-336241698e57.jpg)

I fed a pulse train very similar to F08A (400Hz, 9.88us pulse width) into the F09 scaler stage and captured most of the interesting signals with a logic analyzer.
![waveforms](https://cloud.githubusercontent.com/assets/94056/22583960/a8ec9478-e9a4-11e6-8f6d-690b1ce376a0.jpg)

The expected relationship between TIME3 and TIME4 holds; T4P is pulsing 2.5ms before T3P. Furthermore, the observations in the simulator are still valid -- the order of pulses is:
1. F10A
2. T4P
3. F10B

In order for Ed's loop to always work, that pulse order would need to be
1. F10B
2. T4P
3. F10A

So, at the very least, the timer pulse behavior is not a simulation artifact. Unfortunately, the TC0 and TCF0 transients would require having almost a whole AGC wired up to observe in the real world, so that's going to have to wait for a while.

**Comment by thewonderidiot** [comment]
I'm digging a bit further into the exact conditions for an incorrectly reset TC Trap in anticipation for soon getting proof one way or the other. I've expanded a bit on the causes behind "1. Slightly early clearing of extend for BZF leads to a TCF0 transient" in the list above. The precise definition for that transient source is:

Any BZF that gets taken, for which the target instruction is not EXTEND, causes a long TCF0 transient.

In other words, BZFs executed when A is not +-0, or a BZF that branches to an EXTEND instruction, will not reset the TC Trap. This is significant because of the following sequence:
```
        CA   ZERO
LOOP    EXTEND
        BZF  LOOP
```

This is the most obvious, simple loop you can make if you're trying to trigger the "No TC" part of the TC Trap. And importantly, it actually works, because even though the BZF does get taken, the target instruction is EXTEND so no transients occur. I'm willing to bet that NEWSPEAK, LAMESH, and any other factory test ropes did in fact include exactly this sequence of instructions, which led the designers to incorrectly believe that the TC Trap was working as intended.

**Comment by thewonderidiot** [comment]
One last experiment for now: I rewired the TC Trap a bit in the hardware sim to reflect what I believe it *should* have looked like:

![tctrap_patched](https://cloud.githubusercontent.com/assets/94056/24592456/1c8329a6-17cc-11e7-95d6-bc2e5bfa0790.png)

The only change from the original circuit is that the "No TCs" flip-flop can only be reset during timepulse 4, similar to the "Only TCs" half. This guarantees that all of the transients (which all occur during timepulse 12) are rejected, and only true instructions are considered. Running the Aurora self-tests with this hardware configuration causes a TC Trap, as predicted by VirtualAGC.

**Comment by thewonderidiot** [comment]
This investigation was continued and "resolved", as much as it probably can be, in #805, so go there for a final writeup. The executive summary is that we got a new manual that confirmed all of the above timing findings, so it is quite likely that this was a real bug.
